### PR TITLE
Prompt an error page on account locked situations

### LINKED ISSUE #8600: Show account locked message upon exceeding maximum unsuccessful attempts - Email OTP
**Describe the issue:**
After exceeding the maximum number of unsuccessful OTP attempts, system doesn't display "account is locked" screen or error page, it instead shows a page to enter security code again.

**How to reproduce:**
Configure email OTP as a second factor authenticator.
Enable account locking under resident IDP -> login policies -> account locking.
Try the authentication flow and enter incorrect email OTPs until the maximum number of attempts exceed.

**Expected behavior:**
After the maximum number of attempts are exceeded, the user must be prompted with an error page.

### PR DESCRIPTION
**Pull request by AnuradhaSK**
## Purpose
Resolves https://github.com/wso2/product-is/issues/8600

## Related PRs
https://github.com/wso2-support/identity-apps/pull/17

### CODE REVIEW

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 1]
```diff
@@ -2389,4 +2410,29 @@ private void setUserClaimValues(AuthenticatedUser authenticatedUser, Map<String,
...
+        String username = authenticatedUser.toFullQualifiedUsername();
+        String tenantAwareUsername = MultitenantUtils.getTenantAwareUsername(username);
+        String unlockTime = null;
+        try {
+            UserRealm userRealm = getUserRealm(username);
+            if (userRealm == null) {
+                throw new AuthenticationFailedException("UserRealm is null for user : " + username);
+            }
+            UserStoreManager userStoreManager = userRealm.getUserStoreManager();
+            if (userStoreManager != null) {
```
Comment (somindatommy) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: Shall we change this logic to check `null` and add a debug log?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 2]
```diff
@@ -887,9 +887,30 @@ private void redirectToErrorPage(HttpServletResponse response, AuthenticationCon
...
         Map<String, String> authenticatorProperties = context.getAuthenticatorProperties();
         Map<String, String> emailOTPParameters = getAuthenticatorConfig().getParameterMap();
+        boolean showAuthFailureReason =
+                Boolean.parseBoolean(emailOTPParameters.get(EmailOTPAuthenticatorConstants.SHOW_AUTH_FAILURE_REASON));
         try {
+            if (isLocalUser(context) &&
+                    EmailOTPUtils.isAccountLocked(getAuthenticatedUser(context))) {
+                String retryParam;
+                if (showAuthFailureReason) {
+                    String unlockTime = getUnlockTime(getAuthenticatedUser(context));
```
Comment (somindatommy) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: Can this return null? If so shall we add a debug log with more context to identify that?

Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: fixed in be18eca

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java (hunk 1 above)
Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: fixed in be18eca

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 3]
```diff
@@ -2389,4 +2408,37 @@ private void setUserClaimValues(AuthenticatedUser authenticatedUser, Map<String,
                     String.format("Failed to update user claims for user : %s.", authenticatedUser), e);
         }
     }
+
+    private long getUnlockTimeInMilliSeconds(AuthenticatedUser authenticatedUser) throws AuthenticationFailedException {
```
Comment (somindatommy) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java:2418]: Shall we add a method comment with the behavior of the method? Why we return `0` likewise? WDYT?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 4]
```diff
@@ -2389,4 +2408,37 @@ private void setUserClaimValues(AuthenticatedUser authenticatedUser, Map<String,
...
+                    log.debug("userStoreManager is null for user realm: " + userRealm);
+                }
+                throw new AuthenticationFailedException("userStoreManager is null for user realm: " + userRealm);
+            }
+            Map<String, String> claimValues = userStoreManager
+                    .getUserClaimValues(tenantAwareUsername,
+                            new String[]{EmailOTPAuthenticatorConstants.ACCOUNT_UNLOCK_TIME_CLAIM}, null);
+            if (claimValues.get(EmailOTPAuthenticatorConstants.ACCOUNT_UNLOCK_TIME_CLAIM) == null) {
+                if (log.isDebugEnabled()) {
+                    log.debug("No value configured for claim: " +
```
Comment (somindatommy) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: Shall we add more context into this log?

**Review by somindatommy — APPROVED** [review]

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java (hunk 4 above)
Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: fixed in b87fff7

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java (hunk 3 above)
Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java:2418]: fixed in b87fff7

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 5]
```diff
@@ -887,9 +887,28 @@ private void redirectToErrorPage(HttpServletResponse response, AuthenticationCon
...
                                      String username, String queryParams,
                                      AuthenticationContext context) throws AuthenticationFailedException {
+
         Map<String, String> authenticatorProperties = context.getAuthenticatorProperties();
         Map<String, String> emailOTPParameters = getAuthenticatorConfig().getParameterMap();
+        boolean showAuthFailureReason =
+                Boolean.parseBoolean(emailOTPParameters.get(EmailOTPAuthenticatorConstants.SHOW_AUTH_FAILURE_REASON));
         try {
+            if (isLocalUser(context) &&
+                    EmailOTPUtils.isAccountLocked(getAuthenticatedUser(context))) {
```
Comment (sachiniWettasinghe) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: Shall we format this to one line? Seems like it's not exceeding maximum char length.

Comment (AnuradhaSK) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]: fixed in a8a936f

**Review by sachiniWettasinghe — APPROVED** [review]

**Pull request merged**